benches/pcs.rs

```diff
@@ -0,0 +1,181 @@
+  eval: <E as Engine>::Scalar,
+  ck: <<E as Engine>::CE as CommitmentEngineTrait<E>>::CommitmentKey,
+  commitment: <<E as Engine>::CE as CommitmentEngineTrait<E>>::Commitment,
+  prover_key: <EE as EvaluationEngineTrait<E>>::ProverKey,
+  verifier_key: <EE as EvaluationEngineTrait<E>>::VerifierKey,
+  proof: Option<<EE as EvaluationEngineTrait<E>>::EvaluationArgument>,
+}
+
+// Macro to generate benchmark code for multiple engine and evaluation engine types
+macro_rules! benchmark_engines {
+    ($ell:expr, $rng:expr, $group:expr, $internal_fn:expr, $( ($engine:ty, $eval_engine:ty, $engine_name:expr) ),*) => {
+        $(
+            let mut assets = deterministic_assets::<$engine, $eval_engine, StdRng>($ell, &mut $rng);
+            $group.bench_with_input(BenchmarkId::new($engine_name, $ell), &$ell, |b, _| {
+                $internal_fn(b, &mut assets);
+            });
+        )*
+    };
+}
+
+fn bench_pcs(c: &mut Criterion) {
+  for &ell in TEST_ELL.iter() {
+    let mut rng = rand::rngs::StdRng::seed_from_u64(ell as u64);
+
+    // Proving group
+    {
+      let mut proving_group = c.benchmark_group("PCS-Proving");
+      proving_group
+        .sampling_mode(SamplingMode::Flat)
+        .sample_size(10);
+
+      benchmark_engines!(
+          ell,
+          rng,
+          proving_group,
+          bench_pcs_proving_internal,
+          (GrumpkinEngine, IPAEvaluationEngine<GrumpkinEngine>, "IPA"),
+          (Bn256EngineKZG, MLEvaluationEngine<Bn256, Bn256EngineKZG>, "MLKZG"),
+          (Bn256EngineZM, ZMPCS<Bn256, Bn256EngineZM>, "ZM")
+      );
+
+      proving_group.finish();
+    }
+
+    // Verifying group
+    {
+      let mut verifying_group = c.benchmark_group("PCS-Verifying");
+      verifying_group
+        .sampling_mode(SamplingMode::Flat)
+        .sample_size(10);
+
+      benchmark_engines!(
+          ell,
+          rng,
+          verifying_group,
+          bench_pcs_verifying_internal,
+          (GrumpkinEngine, IPAEvaluationEngine<GrumpkinEngine>, "IPA"),
+          (Bn256EngineKZG, MLEvaluationEngine<Bn256, Bn256EngineKZG>, "MLKZG"),
+          (Bn256EngineZM, ZMPCS<Bn256, Bn256EngineZM>, "ZM")
+      );
+
+      verifying_group.finish();
+    }
+  }
+}
+
+fn deterministic_assets<E: Engine, EE: EvaluationEngineTrait<E>, R: CryptoRng + RngCore>(
+  ell: usize,
+  mut rng: &mut R,
+) -> BenchAssests<E, EE> {
+  let (poly, point, eval) = MultilinearPolynomial::random_with_eval(ell, &mut rng);
+
+  // Mock commitment key.
+  let ck = E::CE::setup(b"test", 1 << ell);
+  // Commits to the provided vector using the provided generators.
+  let commitment = E::CE::commit(&ck, poly.evaluations());
+
+  let (prover_key, verifier_key) = EE::setup(&ck);
+
+  // Generate proof so that we can bench verification.
+  let proof = EE::prove(
+    &ck,
+    &prover_key,
+    &mut E::TE::new(b"TestEval"),
+    &commitment,
+    poly.evaluations(),
+    &point,
+    &eval,
+  )
+  .unwrap();
+
+  BenchAssests {
```

Comment:
BenchAssets?

Reply:
Assets used during our benchmark that are shared between the proving and verification steps. Useful to avoid having lengthy methods signatures.